### ISSUE TITLE
json-c and jansson libraries have symbol conflicts

### ISSUE THREAD

**Issue by vysakh-av** [json-c/json-c#621]
**Describe the bug**
In the application, we are using a shared library which is using json-c for its internal purpose. Later our application required to use a third party library, but its internally used jansson inside it for its functionality. Unfortunately when we run the application , it leads to memory corruption and process crashed.

Both of these libraries have multiple APIs with same name but different signature.
Due to this, there is a memory corruption that resulted in crash.

The following APIs are available in both json-c and jansson libraries with the same name:
json_object_get     
json_object_iter_next

**Steps To Reproduce**
This issue can easily reproduce with the following way:

- `Application` linked to libjson_c_lib.so and libjansson_c_lib.so. Both of these are wrapper libraries are using json-c and jannson internally.

- libjson_c_so wrapper library linked to libjson-c.so
- libjansson_c.so wrapper library linked to libjansson.so.

**Tried statically linked one of the library but its not helped here. If we statically link jansson to wrapper library, json-c calls failed especially json_tokener_parse(). If we statically link json-c library, its resulted in same memory corruption.**

**Comment by hawicz** [comment]
So what do you want us to do?

**Comment by vysakh-av** [comment]
Thank you for the support.
Here do we have any way to resolve this issue, we have tried different ways by statically linking and using -Bsymbolic linker flags, but it didn't help us. Any  help or support much appreciated.

**Comment by micahsnyder** [comment]
This issue ties in with my request to build with `-fPIC` by default, so other libraries can link with the static libjson-c.a library.  Without `-fPIC`, you can only statically link with an application, meaning you can't rely on json-c as a dependency for another library because a downstream application may choose jansson and then crash with symbol collisions.

See my ticket for more details https://github.com/json-c/json-c/issues/508

**Comment by smcv** [comment]
The possible ways to resolve this conflict go something like this:

* json-c renames its symbols to something like `jsonc_whatever()`, which is an API and ABI break, and presumably the json-c maintainers would prefer not to do it
* jansson renames its symbols to something like `jansson_whatever()`, which is an API and ABI break, and presumably the jansson maintainers would prefer not to do it
* both libraries rename their symbols, as above
* *both* libraries use [versioned symbols](https://sourceware.org/binutils/docs/ld/VERSION.html), at least on Linux, and then distributions rebuild dependent binaries against the release that has versioned symbols

Using versioned symbols would mean that:

* when dependent binaries that were compiled against a new enough shared json-c call `json_object_get()`, it's really a reference to `json_object_get@JSON_C` or similar;
* when dependent binaries that were compiled against a new enough shared jansson call `json_object_get()`, it's really a reference to `json_object_get@JANSSON` or similar;
* at runtime, `ld.so` chooses the right implementation of `json_object_get` in each case;
* dependent binaries that have not yet been recompiled against a new enough json-c or jansson still have a reference to plain `json_object_get`, which can be satisfied by either of the libraries, so they will continue to get the wrong one and crash about half the time
* dependent binaries that were statically linked to either json-c or jansson still have a reference to plain `json_object_get`, and will potentially still crash unless they are linked with `-Wl,-Bsymbolic`

None of these solutions are perfect: renaming symbols is an ABI break, while adding versioned symbols requires action from the maintainers of *both* libraries.

If the upstream maintainers of these libraries don't take action, it is possible that downstream maintainers in Linux distributions will decide to add versioned symbols without coordination with upstream. This is usually a bad idea and leads to binary-compatibility problems in the future, and I would recommend that downstream maintainers shouldn't do that, but they might consider it to be a necessary evil to fix crashes.

**Comment by smcv** [comment]
GNOME's json-glib has a similar issue, and in fact all three libraries (json-c, jansson, json-glib) have a `json_object_iter_next()`.

The easiest way to have versioned symbols is: on supported platforms (at least GNU/Linux) and with supported compilers (at least gcc and clang), link the shared library with `-Wl,--default-symver`. This decorates every symbol with a symbol-version that is identical to the SONAME. For example, `json_object_get` becomes `json_object_get@libjson-c.so.5`.

This simple form of versioned symbols is more or less equivalent to `-Wl,--version-script=json-c.ver` where `json-c.ver` has these contents:

```
libjson-c.so.5 {
  global:
    *;
};
```

Either of those is enough to make newly-compiled json-c-dependent objects find the json-c version of a function in preference to other versions. It will not prevent json-glib-dependent or libjansson-dependent objects from crashing when they find json-c's version of a symbol instead of the one they wanted: to prevent that, json-glib and libjansson would need to adopt versioned symbols as well.

Adding versioned symbols is a compatible change and does not need a SONAME bump, but if you adopt versioned symbols, they will become part of the library's ABI: removing or changing them is an incompatible change which requires a SONAME bump.

It is also possible to use more complicated version scripts that assign versions to individual symbols, as is done in projects like OpenSSL, util-linux/libmount/libblkid and libgcab. This requires maintaining or generating a detailed list of symbols, and can help downstream projects to track what the library's dependencies are, but is not necessary if all you want to achieve is to resolve the symbol clash between json-c, json-glib and libjansson.

**Comment by karelzak** [comment]
Guys, symbols versioning is the very basic thing you need in arbitrary serious shared library. Really. 

It's the way how to maintain backward compatibility, avoid soname changes, and how to avoid symbols conflicts. And it's very very simple, all you need is one file. 

Example:
    https://github.com/karelzak/util-linux/blob/master/libmount/src/libmount.sym

You also want to read Ulrich's excellent shared libs how-to (3.3. ABI versioning):
   https://software.intel.com/sites/default/files/m/a/1/e/dsohowto.pdf

**Comment by karelzak** [comment]
@smcv it would be better to maintain symbols file and explicit prefix (see an example with libmount). The default --default-symver is poor man solution.

**Comment by hawicz** [comment]
Using versioned symbols sounds like the way to go.  Using `--default-symver` seems like it would cause non-backwards-compatible ABI changes to 100% of the symbols as soon as the soname changes, so I'm inclined to go with a version script with an explicit list of symbols.  
Is there an easy way to check that you're not missing symbols from your version script?

**Comment by micahsnyder** [comment]
Symbol versioning sounds nice.  Does it work on macOS, other Unix variants, and Windows or would this fix be for Linux/ELF only?  It's _probably_ find if it is -- since most Mac/Windows software is bundled with dependencies and they can work around it with static linking, though I reckon some macOS Homebrew enthusiasts may disagree.

**Comment by folknor** [comment]
> Is there an easy way to check that you're not missing symbols from your version script?

I don't know what the best way is, but `nm` seems like it could be useful here. Try this command for example; `nm -D --defined-only --with-symbol-versions /lib/x86_64-linux-gnu/libmount.so.1`.

The output it generates seems at first glance to match https://github.com/karelzak/util-linux/blob/master/libmount/src/libmount.sym, and writing a script to verify this seems trivial.

**Comment by smcv** [comment]
> Using --default-symver seems like it would cause non-backwards-compatible ABI changes to 100% of the symbols as soon as the soname changes

The purpose of the SONAME is that it changes if and only if you have non-backwards-compatible ABI changes, so that's a feature more than a bug.

If json-c has non-backwards-compatible ABI changes, then in addition to bumping the SONAME, it *should* change the version on all its symbols (either automatically via `--default-symver`, or via maintainer action editing an explicit linker script that resembles the one in libmount). Otherwise, you can get crashes similar to the ones described in this issue, except that instead of the collision being between json-c and jansson, it's between old json-c and new json-c.

**Comment by hawicz** [comment]
Why would we change the version on _all_ symbols if only _some_ of the ABI changes?  e.g. let's say we the only breaking change we make is to rework json_tokener to keep its internal state private: in that case I don't see any reason to change the version on interfaces from json_object.h which stay exactly the same.
Anyway, those details can be sorted out whenever we do have a soname bump.

**Comment by hawicz** [comment]
@folknor, thanks for the suggestion.  It seems my nm doesn't support `--with-symbol-versions`, so I can't rely on that, but even without that the number of "extra" symbols in the output (e.g. _init, _fini, etc...) seems small enough to easily filter out, so writing a test that cross-checks the version script probably wouldn't be too hard.

**Comment by smcv** [comment]
> Why would we change the version on *all* symbols if only *some* of the ABI changes?

tl;dr: Because changing it won't cause new crashes, and might prevent some, depending what else happens. Not changing it doesn't have any benefit, and might cause crashes later, even if it doesn't immediately.

For example, if you take a look at OpenSSL 1.1's `libcrypto.so.1.1` with `objdump -Tx`, you'll see that its implementation of `SHA1_Init()` is versioned `SHA1_Init@OPENSSL_1_1_0`, even though I'm fairly sure the data structure used to compute SHA1 has not changed since much earlier OpenSSL versions.

The reason I say that there is no benefit to *not* changing the symbol-version is that if the SONAME has changed, then binaries linked to the old version (with `DT_NEEDED: libjson-c.so.4`) and binaries linked to the new version (with `DT_NEEDED: libjson-c.so.5`) are already different, and already have a dependency on the SONAME they were linked against, even if the subset of the json-c API they call into happens to be in the subset that is identical between the two versions (for example, if you break the ABI for the tokener but not for `json_object` like you said, and these two json-c users happen to be only using the `json_object` interfaces).

If the SONAME has changed, and a program has one dependency (let's say libfoo) that was built against `libjson-c.so.4` and one (let's say libbar) that was built against `libjson-c.so.5`, then they're both going to end up in the program's address space anyway (until libfoo is rebuilt against the new json-c), because of how SONAMEs and `ld.so` work. If the version on each symbol is different, then libfoo will continue to use the `libjson-c.so.4` copy, libbar will continue to use the `libjson-c.so.5` copy, and everything continues to work as expected.

However, if the `libjson-c.so.4` copies of other `json_object` functions have the same name and symbol-version as the `libjson-c.so.5` copies, then you cannot control which one the linker will choose: it might be either one. If the internal data structures that implement `json_object.h` happen to be identical in both, then yes, maybe it would be safe for both libfoo and libbar to call an arbitrarily-chosen copy of `json_object.h` functions. However, as soon as `libjson-c.so.5` changes anything about the internal data structures - perhaps adding a new struct member to support new functionality, or reducing the struct size because fields were packed more efficiently - they would become incompatible with the `libjson-c.so.4` data structures, and you're back to having it crash if an inconsistent mixture of copies of these functions is chosen. Changing the version attached to a symbol is an ABI break (because the library previously provided `symbol@version1` and now it only provides `symbol@version2`), so you can't escape from that situation by changing the symbols' versions either; the only solution would be another SONAME break.

As long as they have exactly the same set of symbols acting on the same data structures, maybe it'll work. However, if `libjson-c.so.5` adds a `json_object_new_array2()` that didn't exist in `libjson-c.so.4`, then callers of `json_object_new_array2()` will definitely use the `libjson-c.so.5` copy, because there is no `libjson-c.so.4` copy available - but they might still be using the `libjson-c.so.4` copies of other functions that act on the same data structure. If the internal data structure differs from `libjson-c.so.4`, then the `libjson-c.so.4` copies of the other `json_object` functions are going to crash when they try to access it.

**Comment by hawicz** [comment]
Thanks for the explanation, that makes sense.  I included a mention of this in RELEASE_NOTES.txt
With the most recent few commits we now have a version script in place, so I think that's about the most we can do.
